Hyper/Controllers/SkyboxController.cs

```diff
@@ -0,0 +1,229 @@
+        }
+
+        if (time >= sunsetStartTime && time < duskStartTime)
+        {
+            return (_sunsetStart, _duskStart, GetT(sunsetStartTime, duskStartTime, time));
+        }
+
+        if (time >= duskStartTime && time < nightStartTime)
+        {
+            return (_duskStart, _nightStart, GetT(duskStartTime, nightStartTime, time));
+        }
+        if (time >= nightStartTime && time < nightEndTime)
+        {
+            return (_nightStart, _nightStart, 1);
+        }
+        if (time >= nightEndTime && time < dawnStartTime)
+        {
+            return (_nightStart, _dawnStart, GetT(nightEndTime, dawnStartTime, time));
+        }
+
+        return (_dawnStart, _sunriseStart, GetT(dawnStartTime, 2 * _dayLengthSeconds, time));
+
+        static float GetT(float begin, float end, float time)
+        {
+            return (time - begin) / (end - begin);
+        }
+    }
+
+    public void Render()
+    {
+        _skyboxShader.SetUp(_scene.Camera);
+        float time = GetCurrentTime();
+
+        var interval = GetInterval(time);
+        _skyboxShader.SetStruct("prevPhase", interval.Item1);
+        _skyboxShader.SetStruct("nextPhase", interval.Item2);
+        _skyboxShader.SetFloat("phaseT", interval.Item3);
+
+        _skybox.RotationX = time / _dayLengthSeconds * MathF.PI;
+        var sunVector = _initialSunVector * Matrix3.CreateRotationX(_skybox.RotationX);
+        var moonVector = _initialMoonVector * Matrix3.CreateRotationZ(_skybox.RotationX);
+        DirectionalLight sunLight = new DirectionalLight
+        {
+            Ambient = 0.1f,
+            Diffuse = 0.8f,
+            Specular = 0.1f,
+            Direction = new Vector4(sunVector, 0)
+        };
+
+        DirectionalLight moonLight = new DirectionalLight
+        {
+            Ambient = 0.05f,
+            Diffuse = 0.2f,
+            Specular = 0.05f,
+            Direction = new Vector4(moonVector, 0)
+        };
+
+        EnvironmentInfo envInfo = new EnvironmentInfo
+        {
+            PrevPhaseSunLightColor = interval.Item1.SunglareColor.Xyz * (1 - interval.Item1.StarsVisibility) * (1 - interval.Item1.StarsVisibility),
+            NextPhaseSunLightColor = interval.Item2.SunglareColor.Xyz * (1 - interval.Item2.StarsVisibility) * (1 - interval.Item2.StarsVisibility),
+            PrevPhaseMoonLightColor = new Vector3(0.55f, 0.62f, 0.79f) * interval.Item1.StarsVisibility,
+            NextPhaseMoonLightColor = new Vector3(0.55f, 0.62f, 0.79f) * interval.Item2.StarsVisibility,
+            PhaseT = interval.Item3,
+            PrevPhaseNightAmbient = 0.05f * interval.Item1.StarsVisibility,
+            NextPhaseNightAmbient = 0.05f * interval.Item2.StarsVisibility,
+        };
+
+        _modelShader.SetBool("hasSun", true);
+        _modelShader.SetStruct("sunLight", sunLight);
+        _modelShader.SetStruct("moonLight", moonLight);
+        _modelShader.SetStruct("envInfo", envInfo);
+
+        _objectShader.SetBool("hasSun", true);
```

Comment:
Is my understanding corrent that we set up the shaders to later use tham in other controllers> If so I really dont like it because we might have some bugs in another controller without realizing that this controller is the one causing it. I think controllers should communicate through scene.cs and if that doesn't work maybe create some extra non controller that sets this stuff up

Reply:
I honestly don't know what to think about this, it surely looks weird and the same problem is present in other places, like the teleporting logic hidden in player and car controllers

Reply:
But at least as far as i know the teleporter in player controller only does stuff to the player so the hiding isn't that bad. Here we change something completely outside of this controllers responsibility ie Skybox